src/libsyntax/tokenstream.rs

```diff
@@ -417,112 +462,172 @@ fn covering_span(trees: &[TokenTree]) -> Span {
     }
 
     /// Manually change a TokenStream's span.
     pub fn respan(self, span: Span) -> TokenStream {
-        TokenStream {
-            tts: self.tts,
-            span: span,
-        }
+      match self {
+        TokenStream::Empty => self,
+        TokenStream::Leaf { tts, .. } => TokenStream::Leaf { tts : tts, sp : span },
+        TokenStream::SubLeaf { tts, offset, len, .. } =>
+          TokenStream::SubLeaf { tts : tts, offset: offset, len: len, sp : span },
+        TokenStream::Node { left , right, len, .. } =>
+          TokenStream::Node { left: left, right: right, len: len, sp: span },
+      }
     }
 
-    /// Construct a TokenStream from an ast literal.
-    pub fn from_ast_lit_str(lit: ast::Lit) -> Option<TokenStream> {
-        match lit.node {
-            LitKind::Str(val, _) => {
-                let val = TokLit::Str_(token::intern(&val));
-                Some(TokenStream::from_tts(vec![TokenTree::Token(lit.span,
-                                                                 Token::Literal(val, None))]))
-            }
-            _ => None,
-        }
-
+    /// Copies all of the TokenTrees from the TokenSlice, appending them to the stream.
+    pub fn concat(left: TokenStream, right: TokenStream) -> TokenStream {
+      let new_len = left.len() + right.len();
+      let new_span = combine_spans(left.span(), right.span());
+      TokenStream::Node{
+        left : Rc::new(left),
+        right: Rc::new(right),
+        len: new_len,
+        sp: new_span,
+      }
+    }
+
+    /// Indicate if the TokenStream is empty.
+    pub fn is_empty(&self) -> bool {
+      self.len() == 0
     }
 
-    /// Convert a vector of TokenTrees into a parentheses-delimited TokenStream.
-    pub fn as_paren_delimited_stream(tts: Vec<TokenTree>) -> TokenStream {
-        let new_sp = covering_span(&tts);
-
-        let new_delim = Rc::new(Delimited {
-            delim: token::DelimToken::Paren,
-            open_span: DUMMY_SP,
-            tts: tts,
-            close_span: DUMMY_SP,
-        });
-
-        TokenStream::from_tts(vec![TokenTree::Delimited(new_sp, new_delim)])
+    /// Return a TokenSlice's length.
+    pub fn len(&self) -> usize {
+      match *self {
+        TokenStream::Empty => 0,
+        TokenStream::Leaf { ref tts, .. } => tts.len(),
+        TokenStream::SubLeaf { len, .. } => len,
+        TokenStream::Node { len, .. } => len,
+      }
+    }
+
+    /// Convert a TokenStream into a vector of borrowed TokenTrees.
+    pub fn to_vec(&self) -> Vec<&TokenTree> {
+      match *self {
+        TokenStream::Empty => Vec::new(),
+        TokenStream::Leaf { ref tts, .. } => {
+          let mut v = Vec::new();
+          for t in tts.iter() {
+            v.push(t);
+          }
+          v
+        }
+        TokenStream::SubLeaf { ref tts, offset, len, .. } => {
+          let mut v = Vec::new();
+          for i in offset..(offset+len-1) {
+            v.push(tts.get(i).unwrap())
+          }
+          v
```

Comment:
Nit: this could be rewritten as:

```
(offset .. (offset + len - 1))
    .map(|i| tts.get(i).unwrap())
    .collect()
```

Also, could you replace the `.unwrap()` with an `.expect(...)` that gives a little more information about what should have happened so we get a better panic message?